AppDB/appscale/datastore/cassandra_env/large_batch.py

```diff
@@ -214,3 +224,214 @@ def claim(self):
+      VALUES (?, ?, ?)
+      IF NOT EXISTS
+    """)
+    bound_statement = statement.bind((txid_hash, False, op_id))
+    bound_statement.retry_policy = NO_RETRIES
+    results = yield self._tornado_cassandra.execute(bound_statement)
+    if not results[0].applied:
+      raise BatchNotOwned('Another process started applying the transaction')
+
+  @gen.coroutine
+  def _select_mutations(self, txid):
+    """ Fetches a list of the mutations for the batch.
+
+    Args:
+      txid: An integer specifying a transaction ID.
+    Returns:
+      An iterator of Cassandra results.
+    """
+    statement = self._get_prepared("""
+      SELECT old_value, new_value FROM batches
+      WHERE app = ? AND transaction = ?
+    """)
+    bound_statement = statement.bind((self.project_id, txid))
+    bound_statement.retry_policy = BASIC_RETRIES
+    results = yield self._tornado_cassandra.execute(bound_statement)
+    raise gen.Return(results)
+
+  @gen.coroutine
+  def _apply_mutations(self, txid, composite_indexes):
+    """ Applies all the mutations in the batch.
+
+    Args:
+      txid: An integer specifying a transaction ID.
+      composite_indexes: A list of CompositeIndex objects.
+    """
+    results = yield self._select_mutations(txid)
+    futures = []
+    for result in results:
+      old_entity = result.old_value
+      if old_entity is not None:
+        old_entity = entity_pb.EntityProto(old_entity)
+
+      new_entity = result.new_value
+
+      if new_entity is None:
+        mutations = deletions_for_entity(old_entity, composite_indexes)
+      else:
+        new_entity = entity_pb.EntityProto(new_entity)
+        mutations = mutations_for_entity(new_entity, txid, old_entity,
+                                         composite_indexes)
+
+      statements_and_params = self._db_access.statements_for_mutations(
+        mutations, txid)
+      for statement, params in statements_and_params:
+        futures.append(self._tornado_cassandra.execute(statement, params))
+
+    for future in futures:
+      yield future
+
+  @gen.coroutine
+  def _update_op_id(self, txid_hash, applied_status, old_op_id, new_op_id):
+    """ Claims a batch that is in progress.
+
+    Args:
+      txid_hash: A byte array identifying the transaction.
+      applied_status: A boolean indicating that the batch has been committed.
+      old_op_id: A uuid4 specifying the last read process ID.
+      new_op_id: A uuid4 specifying the new process ID.
+    """
+    statement = self._get_prepared("""
+      UPDATE batch_status
+      SET op_id = ?
+      WHERE txid_hash = ?
+      IF op_id = ?
+      AND applied = ?
+    """)
+    params = (new_op_id, txid_hash, old_op_id, applied_status)
+    bound_statement = statement.bind(params)
+    bound_statement.retry_policy = NO_RETRIES
+    results = yield self._tornado_cassandra.execute(bound_statement)
+    if not results[0].applied:
+      raise BatchNotOwned('Batch status changed after checking')
```

Comment:
Is there any other possible reason for the result to be not applied? Or BatchNowOwned is one possible cause?

Reply:
If the statement was not applied, it could only be because the op_id changed (indicating another process claimed ownership of the batch) or the applied status changed since reading it (shouldn't be possible if the op_id is the same, but would indicate that another process modified the batch).